find_similar.py

```diff
@@ -0,0 +1,155 @@
+
+def get_move(line):
+    line = str(line)
+    game = []  # list of moves in pgn notation
+    for move in re.findall(
+        r"[^{\[.}\]]+ ", line.replace("?", "").replace("!", "")
+    ):  # Extract all moves without comments
+        game.extend(
+            move.strip().split(" ")
+        )  # Sometimes one string contains two moves with space between
+    return game
+
+
+def find_lowest(array, num, key=key):
+    """
+    Finds x smallest values ​​in a table.
+    :param array: list of records
+    :param num: number of records to be found
+    :param key: sort function key
+    :return: list of index and records
+    """
+    result = []
+    for i, a in enumerate(array):
+        if len(result) < num:
+            result.append((i, a))
+            result.sort(key=key)
+        else:
+            if result[-1][1] > a:
+                result[-1] = (i, a)
+                result.sort(key=key)
+    return result
+
+
+def find_game(idx, cursor):
+    result = []
+    for i in idx:
+        data = cursor.execute(
+            "SELECT Embeding.Autor, Embeding.Number, Embeding.Move FROM Embeding WHERE (((Embeding.Identyfikator)="
+            + str(i[0])
+            + "));"
+        ).fetchall()[0]
+        move = data[2]
+        data = cursor.execute(
+            "SELECT * FROM Games_lite WHERE (((Games_lite.Autor)='"
+            + data[0]
+            + "') AND ((Games_lite.Number)="
+            + str(data[1])
+            + "));"
+        ).fetchall()
+        result.append(list(data[0]) + [move])
+    return result
+
+
+def find_similar(fen, num=1,similarity_function = nearest):
+    """
+    Finds x smallest values ​​in a table.
+    :param fen: String with chess position in Fen notation
+    :param num: Number of games to be found
+    :param similarity_function: Function that measures the similarity of vectors
+    :return: list of games with similar positions
+    """
+    coder = model.Coder(settings.BOARD_SHAPE, settings.LATENT_SIZE).to(settings.DEVICE)
+    # coder.load_state_dict(torch.load(setting.CODER_PATH))
+    coder.eval()
+    inf = Inference(settings.DEVICE, coder)
+    target = inf.predict([fen]).tolist()[0]
+    conn = pyodbc.connect(settings.DATABASE)
+    cursor = conn.cursor()
+    cursor.execute("select Embeding FROM Embeding")
+    matrix = cursor.fetchall()
+    matrix = [json.loads(x[0])[0] for x in matrix]
+    scores = similarity_function(matrix,target)
+    idx = find_lowest(scores, num)
+    return games(find_game(idx, cursor))
+
+
+class games:
+    def __init__(self, data):
+        """
+        :param data: list from find_similar function
+        """
+        self.board = chess.Board()
+        self.moves = []
+        self.current_move = 0
+        self.current_game = 0
+        self.main_move = []
+        self.info = []
+        for game in data:
+            info = {}
+            for inf in game[2].split("|")[:-1]:

```

Comment:
the 5 lines from here were hard to understand for me, I'd like some examples (or maybe, if `data` is a list of tuples, use `for variablename1, variablename2, ... variablenamen in data` that will extract `data` items into variables with meaningful names?`

Reply:
this data format is reading from a database